src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java

```diff
@@ -316,40 +333,141 @@ public void insertBufferedRows() {
+
+    private InsertRowsApiResponseSupplier(
+        SnowflakeStreamingIngestChannel channel, StreamingBuffer streamingBuffer) {
+      this.channel = channel;
+      this.streamingBuffer = streamingBuffer;
+    }
+
+    @Override
+    public InsertValidationResponse get() throws Throwable {
+      LOGGER.debug(
+          "Invoking insertRows API for channel:{}, noOfRecords:{}, startOffset:{}, endOffset:{}",
+          this.channel.getFullyQualifiedName(),
+          streamingBuffer.getNumOfRecords(),
+          streamingBuffer.getFirstOffset(),
+          streamingBuffer.getLastOffset());
+      return this.channel.insertRows(
+          streamingBuffer.getData(), Long.toString(streamingBuffer.getLastOffset()));
+    }
+  }
+
+  /** We will reopen the channel on {@link SFException} and reset offset in kafka. */
+  private void insertRowsFallbackSupplier(StreamingBuffer streamingBuffer) {
+    final long offsetRecoveredFromSnowflake =
+        streamingApiFallbackSupplier(StreamingApiFallbackInvoker.INSERT_ROWS_FALLBACK);
+    LOGGER.warn(
+        "{} Fetched offsetToken:{} for channel:{}",
+        StreamingApiFallbackInvoker.INSERT_ROWS_FALLBACK,
+        offsetRecoveredFromSnowflake,
+        this.getChannelName());
+    throw new RetriableException(
+        String.format(
+            "%s Failed to insert rows for channel:%s, startOffset:%s, endOffset:%s",
+            StreamingApiFallbackInvoker.INSERT_ROWS_FALLBACK,
+            this.getChannelName(),
+            streamingBuffer.getFirstOffset(),
+            streamingBuffer.getLastOffset()));
+  }
+
+  /**
+   * Invoked only when {@link InsertValidationResponse} has errors.
+   *
+   * <p>This function checks if we need to log errors, send it to DLQ or just ignore and throw
+   * exception.
+   *
+   * @param insertErrors errors from validation response. (Only if it has errors)
+   * @param insertedRecordsToBuffer to map {@link SinkRecord} with insertErrors
+   */
+  private void handleInsertRowsFailures(
+      List<InsertValidationResponse.InsertError> insertErrors,
+      List<SinkRecord> insertedRecordsToBuffer) {
+    if (logErrors) {
+      for (InsertValidationResponse.InsertError insertError : insertErrors) {
+        LOGGER.error("Insert Row Error message", insertError.getException());
+      }
+    }
+    if (errorTolerance) {
+      if (!isDLQTopicSet) {
+        LOGGER.warn(
+            "Although config:{} is set, Dead Letter Queue topic:{} is not set.",
+            ERRORS_TOLERANCE_CONFIG,
+            ERRORS_DEAD_LETTER_QUEUE_TOPIC_NAME_CONFIG);
+      } else {
+        for (int i = 0; i < insertErrors.size(); i++) {
+          // Write now I am assuming each error maps to its original row which is incorrect.
+          // TODO: SNOW-545729 would handle mapping errors with original SinkRecord.
+          this.kafkaRecordErrorReporter.reportError(
+              insertedRecordsToBuffer.get(i), insertErrors.get(i).getException());
```

Comment:
I think you can update this now since you're using the new beta release

Reply:
Updated.